src/libsyntax_ext/concat.rs

```diff
@@ -61,6 +73,36 @@ pub fn expand_syntax_ext(
         err.note("only literals (like `\"foo\"`, `42` and `3.14`) can be passed to `concat!()`");
         err.emit();
     }
+    // Do not allow mixing "" and b""
+    if string_accumulator.len() > 0 && b_accumulator.len() > 0 {
+        let mut err = cx.struct_span_err(
+            b_pos.clone(),
+            "cannot concatenate a byte string literal with string literals",
+        );
+        for pos in &b_pos {
+            err.span_label(*pos, "byte string literal");
+        }
+        for pos in &string_pos {
+            err.span_label(*pos, "string literal");
+
+        }
+        err.help("do not mix byte string literals and string literals");
+        err.multipart_suggestion(
+            "you can use byte string literals",
+            string_pos
+                .iter()
+                .map(|pos| (pos.shrink_to_lo(), "b".to_string()))
+                .collect(),
+        );
+        err.emit();
+    }
     let sp = sp.apply_mark(cx.current_expansion.mark);
-    base::MacEager::expr(cx.expr_str(sp, Symbol::intern(&accumulator)))
+    if b_accumulator.len() > 0 {
+        base::MacEager::expr(cx.expr_lit(
+            sp,
+            ast::LitKind::new_byte_str(b_accumulator),
```

Comment:
You might want instead a `cx.expr_byte_str`.

Reply:
done